shared_model/builders/protobuf/builder_templates/query_template.hpp

```diff
@@ -0,0 +1,222 @@
+/**
+ * Copyright Soramitsu Co., Ltd. 2018 All Rights Reserved.
+ * http://soramitsu.co.jp
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef IROHA_PROTO_QUERY_BUILDER_TEMPLATE_HPP
+#define IROHA_PROTO_QUERY_BUILDER_TEMPLATE_HPP
+
+#include "backend/protobuf/queries/proto_query.hpp"
+#include "builders/protobuf/unsigned_proto.hpp"
+#include "interfaces/common_objects/types.hpp"
+#include "interfaces/transaction.hpp"
+#include "queries.pb.h"
+#include "validators/default_validator.hpp"
+
+namespace shared_model {
+  namespace proto {
+
+  /**
+     * Template query builder for creating new types of query builders by
+     * means of replacing template parameters
+     * @tparam S -- field counter for checking that all required fields are set
+     * @tparam SV -- stateless validator called when build method is invoked
+     * @tparam BT -- build type of built object returned by build method
+     */
+    template <int S = 0, typename SV = validation::DefaultQueryValidator, typename BT = UnsignedWrapper<Query>>
+    class TemplateQueryBuilder {
+     private:
+      template <int, typename, typename >
+      friend class TemplateQueryBuilder;
+
+      enum RequiredFields {
+        CreatedTime,
+        CreatorAccountId,
+        QueryField,
+        QueryCounter,
+        TOTAL
+      };
+
+      template <int s>
+      using NextBuilder = TemplateQueryBuilder<S | (1 << s), SV, BT>;
+
+      using ProtoQuery = iroha::protocol::Query;
+
+      template <int Sp>
+      TemplateQueryBuilder(const TemplateQueryBuilder<Sp> &o)
```

Comment:
Probably it should accept only `TemplateQueryBuilder<Sp, SV, BT>`, so that there are no unexpected conversions with different validators or build types.